ocsmesh/driver.py

```diff
@@ -87,69 +98,176 @@ def run(self, sieve=None) -> Mesh:
                 utils.reproject(sizing, calc_crs)
 
-        # Handle Seed CRS mismatch
-        # We assume shape has the "correct" projection (likely UTM from BaseGeom)
-        generation_seed = self._seed_data
+        # 3. Handle 'adapt' mode (Resample Geometry based on Hfun)
+        if self._bnd_representation == 'adapt':
+            if sizing is None:
+                _logger.warning("'adapt' boundary requested but no Hfun provided. Skipping adaptation.")
+            else:
+                _logger.info("Adapting boundary geometry to match Hfun resolution...")
+                shape = self._resample_geom_by_hfun(shape, sizing)
 
+        # 4. Handle Seed CRS
+        generation_seed = self._seed_data
         if generation_seed is not None:
-            # Determine the CRS of the seed data
-            # If explicit CRS is missing, assume it matches the input geometry's CRS
-            # Create a deepcopy to avoid side-effects on the user's original object
             generation_seed = copy.deepcopy(generation_seed)
             seed_crs = generation_seed.crs
-            if calc_crs is None:
-                if seed_crs is not None:
-                    calc_crs = seed_crs
-                    shape = shape.set_crs(calc_crs)
-                    sizing.crs = calc_crs
-            elif seed_crs is None:
-                generation_seed.crs = calc_crs
-            elif not seed_crs.equals(calc_crs):
-                # If reprojection is needed
-
-                _logger.info("Reprojecting seed data to match geometry CRS...")
-                utils.reproject(generation_seed, calc_crs)
-
-        # Pass the prepared seed to generate
+            if seed_crs is not None and not seed_crs.equals(calc_crs):
+                 _logger.info("Reprojecting seed data to match calculation CRS...")
+                 utils.reproject(generation_seed, calc_crs)
+
+        # 5. Generate Mesh
         output_mesh: MeshData = self._engine.generate(
             shape,
             sizing,
             seed=generation_seed
         )
         output_mesh.crs = calc_crs
 
+        # 6. Finalize
         utils.finalize_mesh(output_mesh, sieve)
 
-        if self._crs is not None:
-            if output_mesh.crs is None:
-                output_mesh.crs = self._crs
-            elif not output_mesh.crs.equals(self._crs):
-                utils.reproject(output_mesh, self._crs)
+        # 7. Final Projection
+        target_crs = self._crs
+        if target_crs is None:
+             target_crs = CRS.from_user_input("EPSG:4326")
 
-        return Mesh(output_mesh)
+        if output_mesh.crs is None:
+            output_mesh.crs = target_crs
+        elif not output_mesh.crs.equals(target_crs):
+            _logger.info(f"Reprojecting final mesh to {target_crs}...")
+            utils.reproject(output_mesh, target_crs)
 
+        return Mesh(output_mesh)
 
-    def run_remeshing(self) -> Mesh:
+    def _resample_geom_by_hfun(self, shape_series, hfun_data):
```

Comment:
Is the main reason this is here and not in the engine the CRS handling aspect?

Reply:
Going over the crs logic, I still don't think we need a change in the crs handling here necessarily, and this function can be moved into the engine

Although I see this as being a useful utils too, so we can probably have it in utils and call it in the engine, but not here